### DISCUSSION TITLE
JavascriptObjectRepository.Settings.LegacyBindingEnabled Exception

### DISCUSSION THREAD

**Discussion by Riton2013** [cefsharp/CefSharp#4160] [General]
Cause System.Exception in cefsharp.dll when run to JavascriptObjectRepository.Settings.LegacyBindingEnabled = true;

I use newest version 103.0.9 for winform
my code : 

            chromeBrowser = new ChromiumWebBrowser(Environment.CurrentDirectory.ToString()+"\\index.html");
            
            dSkinPanel1.Controls.Add(chromeBrowser);
            chromeBrowser.Dock = DockStyle.Fill;

            chromeBrowser.JavascriptObjectRepository.Settings.LegacyBindingEnabled = true;
            CefSharpSettings.WcfEnabled = true;
            chromeBrowser.JavascriptObjectRepository.Register("dispatchEvent", new DispatchEvent(), isAsync: true, options: BindingOptions.DefaultBinder);

**Comment by Riton2013** [comment]
ps :  I use .NET 4.8

**Reply by amaitland** [reply]
What is the stack trace? The actual error message? 

The property should ideally be set before the control is added to controls collection.

**Reply by Riton2013** [reply]
Yes, you are right, it should ideally be set before the control is added. I fixed just now, thank you!

**Reply by underwom** [reply]
> The property should ideally be set before the control is added to controls collection.

Thank you for this. I've had some serious struggles getting a JS object to register, and the only relevant documentation I've found says to set the LegacyBindingEnabled property before the object is created. And I have no idea how to set an object's property before that object exists. Moving the register call back to that point as well is the only change that's worked for me.

**Reply by amaitland** [reply]
> the only relevant documentation I've found says to set the LegacyBindingEnabled property before the object is created

What documentation are you referring?

**Reply by underwom** [reply]
Not full docs, just the System.Exception message that shows up only when I try to set LegacyBindingEnabled after adding the browser object to the control. I actually wasn't able to find actual documentation on how or why this works, but that property can be freely set to true up until the point where I add the browser object to the form's control collection.

>JavascriptBindingSettings.LegacyBindingEnabled can no longer be modified, settings must be changed before the underlying browser has been created.

That message sent me on a hunt that eventually led me here, where you commented about that order. Then I compared what I'm doing to your WinForms SimpleBrowserForm example, and ended up making a CreateBrowser method that also handles JS object binding before handing the new browser object back to be added as a form control. The Register method seems to be able to be safely called whenever, but it acts like a no-op if LegacyBindingEnabled is false or if it's called before adding to the control collection. Was there a doc somewhere that I just missed?

**Reply by amaitland** [reply]
> just the System.Exception message that shows up only when I try to set LegacyBindingEnabled after adding the browser object to the control

If you have a suggestion on improving the error message then please take the time to submit a PR.

I read the error and it makes perfect sense, it can be difficult to wrote error messages that everyone understands. 

> The Register method seems to be able to be safely called whenever, but it acts like a no-op if LegacyBindingEnabled is false or if it's called before adding to the control collection. Was there a doc somewhere that I just missed?

Binding by default is initiated from JavaScript, see https://github.com/cefsharp/CefSharp/wiki/General-Usage#binding-an-async-object-in-javascript
